draft-ietf-quic-transport.md

```diff
@@ -1721,7 +1721,12 @@ it can associate the peer's response with the corresponding PATH_CHALLENGE.
 ## Path Validation Responses
 
 On receiving a PATH_CHALLENGE frame, an endpoint MUST respond immediately by
-echoing the data contained in the PATH_CHALLENGE frame in a PATH_RESPONSE frame.
+echoing the data contained in the PATH_CHALLENGE frame in a PATH_RESPONSE frame,
+unless it has PATH_RESPONSE frames buffered for the same destination connection
+ID and wishes to limit memory consumption.
+
+An endpoint MUST NOT store or retransmit a PATH_RESPONSE frame, as a new
+PATH_CHALLENGE frame will be sent if another PATH_RESPONSE frame is needed.
 
```

Comment:
I don't understand this: "unless it has PATH_RESPONSE frame buffered" and later "MUST NOT store"? why would you buffer a response and then not send it, and why would you buffer if you are not allowed to?  And why would it limit memory to buffer and why would it use less memory when choosing not to send the buffered data?

Do you mean "unless a PATH_RESPONSE has already been sent to the same destination connection ID?"

Reply:
I think the original intent here for “store” was in terms of (frames in) packets that are waiting to be acked, not things that might be kept around for a very short time to satisfy pacing and/or congestion limits on their way onto the network. Some of that should be cleaned up soon.

You should send multiple path responses in response to multiple path challenges, even on the same connection ID - unless you’re getting attacked with them at such a high rate that you can’t get them out the door quickly enough and you need to limit your memory consumption while doing that.

Reply:
Maybe something like: "An endpoint MUST NOT send more than one PATH_RESPONSE frame in response to one PATH_REQUEST frame, as a new ..."

Reply:
Why do we bother with ACK on challenge / response? And is that an attack vector, given what you just said?

Reply:
> You should send multiple path responses in response to multiple path challenges, even on the same connection ID

This seems unjustified.  Let's say an endpoint receives 100 _PATH_CHALLENGE_ frames in a single packet.  Is it to reply with 100 _PATH_RESPONSE_ frames?

Reply:
There's no point in putting multiple PATH_RESPONSEs in a single packet, nor in responding to more than one PATH_CHALLENGE from a given packet.  The others can just be ignored.  However, multiple attempts across multiple packets increases resiliency -- if you get 100 PATH_CHALLENGEs in 100 packets, I'd say you send back 100 PATH_RESPONSEs in 100 packets.

Reply:
To back up @MikeBishop point, I'd suggest. "If a packet contains multiple PATH_CHALLENGE frames, only one PATH_RESPONSE should be sent in response."

Reply:
Since there's no point in having multiple challenges in one packet, why not simply forbid it?

Reply:
I'd be fine with forbidding it if we think it's harmful in some way, I'd much prefer that to collapsing multiple PATH_CHALLENGE frames into a single PATH_RESPONSE.

Reply:
I rather prefer the principle where you can process a frame without knowing the rest of the packet content. Since challenges are not retransmitted with the complexity that could entail, it is simpler to send at most one challenge per packet.

Reply:
The issue is not important enough to warrant its own "MUST NOT."  If an endpoint wants to send a packetful of _PATH_CHALLENGE_ frames, let it.  The peer is allowed to send only one _PATH_RESPONSE_.

Reply:
Do we have precedent for a fixed number of something allowed in a packet?  We've previously said multiple ACKs isn't best practice, but it's not prohibited.  This feels similar -- adding a new type of prohibition for something not actively harmful seems excessive.  Simply don't require them to be processed, and let implementations choose.

Reply:
Added text for SHOULD NOT send/reply to the appropriate sections, as with @ianswett and @MikeBishop's suggestions.